### DISCUSSION TITLE
Migrating to Apollo server v4; ValidationError

### DISCUSSION THREAD

**Discussion by alexbjorlig** [apollographql/apollo-server#7093] [Q&A] [answered]
Hi there.

I was looking into upgrading our Apollo server, from v3 to v4. According to the docs [here](https://www.apollographql.com/docs/apollo-server/migration/), it's mentioned that `ValidationError ` is no more.

So my question is, how would you refactor this code from apollo server v3?

```
import { ValidationError } from 'apollo-server';

throw new ValidationError('some-error');
```

Our project uses the above pattern in 144 files 🤣

**Comment by glasser** [comment, marked as answer]
See [this section](https://www.apollographql.com/docs/apollo-server/migration/#built-in-error-classes):

You can replace this with

```
import { GraphQLError } from 'graphql';
import { ApolloServerErrorCode } from '@apollo/server/errors';
throw new GraphQLError("some error", {
  extensions: {
    code:  ApolloServerErrorCode.GRAPHQL_VALIDATION_FAILED',
  },
);
```

Or if you'd like, just define your own class:

```
import { GraphQLError } from 'graphql';
import { ApolloServerErrorCode } from '@apollo/server/errors';
class ValidationError extends GraphQLError {
  constructor(message: string) {
    super(message, { extensions: { code: ApolloServerErrorCode.GRAPHQL_VALIDATION_FAILED } });
  }
}
```

We wanted to move away from the exported error class approach because it led to people trying to do `instanceof ValidationError` checks which could run into trouble if you accidentally ended up with multiple copies of `apollo-server-errors` installed (which was more likely in AS3 with its many-package structure than in AS4 where most code is in `@apollo/server`).

**Reply by alexbjorlig** [reply]
This worked really well, I could do search and replace. Thanks 👍 Probably going to ask a couple of other questions, but very helpful!
